File: libgomp/testsuite/libgomp.fortran/task-in-explicit-1.f90
! { dg-do run }

program main
  use omp_lib
  implicit none
  integer :: i

  if (omp_in_explicit_task ()) &
    error stop
  !$omp task
  if (.not. omp_in_explicit_task ()) &
    error stop
  !$omp end task

  !$omp task final (.true.)
    if (.not. omp_in_explicit_task ()) &
      error stop
    !$omp task
    if (.not. omp_in_explicit_task ()) &
      error stop
    !$omp end task
  !$omp end task

  !$omp parallel
    if (omp_in_explicit_task ()) &
      error stop
    !$omp task if (.false.)
        if (.not. omp_in_explicit_task ()) &
          error stop
        !$omp task if (.false.)
          if (.not. omp_in_explicit_task ()) &
            error stop
        !$omp end task
    !$omp end task
    !$omp task final (.true.)
      if (.not. omp_in_explicit_task ()) &
        error stop
    !$omp end task
    !$omp barrier
    if (omp_in_explicit_task ()) &
      error stop
    !$omp taskloop num_tasks (24)
    do i = 1, 32
      if (.not. omp_in_explicit_task ()) &
        error stop
    end do
    !$omp masked
    !$omp task
    if (.not. omp_in_explicit_task ()) &
      error stop
    !$omp end task
    !$omp end masked
    !$omp barrier
    if (omp_in_explicit_task ()) &
      error stop
  !$omp end parallel

  !$omp target
    if (omp_in_explicit_task ()) &
      error stop
    !$omp task if (.false.)
    if (.not. omp_in_explicit_task ()) &
      error stop
    !$omp end task
    !$omp task
    if (.not. omp_in_explicit_task ()) &
      error stop
    !$omp end task
  !$omp end target

  !$omp target teams
    !$omp distribute
    do i = 1, 4
      if (omp_in_explicit_task ()) then
        error stop
      else
          !$omp parallel
            if (omp_in_explicit_task ()) &
              error stop
            !$omp task
            if (.not. omp_in_explicit_task ()) &
              error stop
            !$omp end task
            !$omp barrier
            if (omp_in_explicit_task ()) &
              error stop
          !$omp end parallel
      end if
    end do
  !$omp end target teams

  !$omp teams
    !$omp distribute
    do i = 1, 4
      if (omp_in_explicit_task ()) then
        error stop
      else
          !$omp parallel
            if (omp_in_explicit_task ()) &
              error stop
            !$omp task
            if (.not. omp_in_explicit_task ()) &
              error stop
            !$omp end task
            !$omp barrier
            if (omp_in_explicit_task ()) &
              error stop
          !$omp end parallel
      end if
    end do
    !$omp end distribute
  !$omp end teams
end

File: libgomp/testsuite/libgomp.c-c++-common/task-in-explicit-1.c
/* { dg-do run } */

#include <omp.h>
#include <stdlib.h>

int
main ()
{
  if (omp_in_explicit_task ())
    abort ();
  #pragma omp task
  if (!omp_in_explicit_task ())
    abort ();
  #pragma omp task final (1)
  {
    if (!omp_in_explicit_task ())
      abort ();
    #pragma omp task 
    if (!omp_in_explicit_task ())
      abort ();
  }
  #pragma omp parallel
  {
    if (omp_in_explicit_task ())
      abort ();
    #pragma omp task if (0)
      {
	if (!omp_in_explicit_task ())
	  abort ();
	#pragma omp task if (0)
	  if (!omp_in_explicit_task ())
	    abort ();
      }
    #pragma omp task final (1)
      if (!omp_in_explicit_task ())
	abort ();
    #pragma omp barrier
    if (omp_in_explicit_task ())
      abort ();
    #pragma omp taskloop num_tasks (24)
    for (int i = 0; i < 32; ++i)
      if (!omp_in_explicit_task ())
	abort ();
    #pragma omp masked
    #pragma omp task
    if (!omp_in_explicit_task ())
      abort ();
    #pragma omp barrier
    if (omp_in_explicit_task ())
      abort ();
  }
  #pragma omp target
  {
    if (omp_in_explicit_task ())
      abort ();
    #pragma omp task if (0)
    if (!omp_in_explicit_task ())
      abort ();
    #pragma omp task
    if (!omp_in_explicit_task ())
      abort ();
  }
  #pragma omp target teams
  {
    #pragma omp distribute
    for (int i = 0; i < 4; ++i)
      if (omp_in_explicit_task ())
	abort ();
      else
	{
	  #pragma omp parallel
	  {
	    if (omp_in_explicit_task ())
	      abort ();
	    #pragma omp task
	    if (!omp_in_explicit_task ())
	      abort ();
	    #pragma omp barrier
	    if (omp_in_explicit_task ())
	      abort ();
	  }
	}
  }
  #pragma omp teams
  {
    #pragma omp distribute
    for (int i = 0; i < 4; ++i)
      if (omp_in_explicit_task ())
	abort ();
      else
	{
	  #pragma omp parallel
	  {
	    if (omp_in_explicit_task ())
	      abort ();
	    #pragma omp task
	    if (!omp_in_explicit_task ())
	      abort ();
	    #pragma omp barrier
	    if (omp_in_explicit_task ())
	      abort ();
	  }
	}
  }
  return 0;
}
